### PR TITLE
AL-Pipeline

### PR DESCRIPTION
**Pull request by paulmorio**
Hi all,

This is a PR related to #23 which we have been planning for a while. As #23 describes we propose refactoring the code to decouple the functionality of the active learning pipeline from the active learning strategy.

In principle the proposed `GenericPipeline` class will succeed the GenericALStrategy in functionality, and serves to help the user decouple the implementation of their strategies from the business logic of running the pipeline.

From a structural perspective a `Pipeline` will be a composition of the following objects

- a DataManager
- a ML model
- an AL strategy
- an oracle (implementing the new `pyrelational.oracle.generic_oracle.GenericOracle` interface)

It retains all of the methods of the previous `GenericActiveLearningStrategy` it replaces. Of particular note is the `active_learning_step` method which now calls upon the method of the same name in the strategy, filtering the arguments to match the signature of the concrete method implementations.

This involves the modification of several existing modules, most notably the `GenericActiveLearningStrategy` and all of the strategy implementations dependent on it. The refactor also includes a new `GenericOracle` interface to spur development of different concrete oracle classes. 

The `GenericActiveLearningStrategy` is now an abstract class which acts as an interface for strategy implementations, each concrete strategy must implement an `active_learning_step(self, *args, **kwargs)` function. The `GenericPipeline.active_learning_step(self, *args, **kwargs)` will filter the `**kwargs` to match signature of the implemented `GenericActiveLearningStrategy.active_learning_step` method. 

The change also brings forward the implementation of a `GenericOracle` interface whose concrete implementations must implement the `update_dataset(self, data_manager, indices)` method.

Updates will be made to this PR to update all of the tests and documentations towards the new API as we discuss any detailed changes that need to be done to the proposed changes.

### CODE REVIEW

**Comment by paulmorio** [comment]
Updated the active_learning_step pattern to input the pipeline attributes as default kwargs that get overwritten/updated by any user kwargs.

Also added a working example (with the associated refactoring needed in the dependent modules)

**Comment by paulmorio** [comment]
All tests updated to reflect new API changes

**Review by a-pouplin — DISMISSED** [review]
Everything seems good to me !

File: pyrelational/pipeline/generic_pipeline.py [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+        oracle: Optional[GenericOracle] = None,
+    ):
+        super(GenericPipeline, self).__init__()
+        self.data_manager = data_manager
+        self.model = model
+        self.strategy = strategy
+        self.oracle = oracle
+
+        # Mediator pattern for strategy
+        self.strategy._pipeline = self
```
[the remark this answers is not in the document]
Comment (thomasgaudelet) [reply pyrelational/pipeline/generic_pipeline.py:59]: not sure this is actually used in the strategies so far as well

File: pyrelational/strategies/regression/bald_strategy.py [hunk 2]
```diff
@@ -27,18 +27,18 @@ class SoftBALDStrategy(BALDStrategy):
...
         assert temperature > 0, "temperature parameter should be greater than 0"
         self.T = torch.tensor(temperature)
 
-    def active_learning_step(self, num_annotate: int) -> List[int]:
-        self.model.train(self.l_loader, self.valid_loader)
-        output = self.model(self.u_loader)
+    def active_learning_step(
+        self, num_annotate: int, data_manager: GenericDataManager, model: GenericModel
+    ) -> List[int]:
+        model.train(data_manager.get_labelled_loader(), data_manager.get_validation_loader())
```
Comment (thomasgaudelet) [inline pyrelational/strategies/regression/bald_strategy.py:39]: maybe, we should brainstorm a way to abstract this away.

**Review by thomasgaudelet — CHANGES_REQUESTED** [review]
A few comments on the surface. Looking forward to chat tomorrow

File: pyrelational/pipeline/generic_pipeline.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (paulmorio) [reply pyrelational/pipeline/generic_pipeline.py:59]: Apologies I initially tried building it using a mediator design pattern, but went in favour of using the current  _filter_kwargs pattern, so this is a deprecated artifact. This will be removed in the next commit.

File: pyrelational/pipeline/generic_pipeline.py [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+        :param num_annotate: number of observations to get annotated per iteration
+        :param num_iterations: number of active learning loop to perform
+        :param oracle_interface: undefined for now, this will be entry point for external oracle later
+        :param test_loader: test data with which we evaluate the current state of the model given the labelled set L
+        :param return_query_history: whether to return the history of queries or not
+        :return: optionally returns a dictionary storing the indices of queries at each iteration
+        """
+        iter_count = 0
+        if return_query_history:
+            query_history = {}
```
[the remark this answers is not in the document]
Comment (paulmorio) [reply pyrelational/pipeline/generic_pipeline.py:184]: Now that you mention it, I think we need a new way to do the logging entirely, as of this moment `query_history` stores the observation id along with the iteration at which it was labelled, whilst the `log_labelled_by` stores the tag. We could supercede both with a better name and data structure which stores both of these things.

File: pyrelational/pipeline/generic_pipeline.py [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+
+        # Final update the model and check final test performance
+        self.model.train(self.l_loader, self.valid_loader)
+        self.performances[self.iteration] = self.current_performance(test_loader=test_loader)
+        if return_query_history:
+            return query_history
+
+    def update_annotations(self, indices: List[int]) -> None:
+        """
+        Update labels in the datamanager train set based on indices selected for labelling
```
[the remark this answers is not in the document]
Comment (paulmorio) [reply pyrelational/pipeline/generic_pipeline.py:217]: Ok I will implement a default oracle which simply passes the `update_dataset` method defined in `oracle.generic_oracle.GenericOracle`, this should take care of the if statement within the generic_pipeline.

As for the first part, I will simply provide a few more details in the docstring. The `update_annotations` method simply calls `data_manager.update_train_labels(indices)` which simply shifts the relevant unlabelled indices to the labelled indices. Provided that the `update_dataset` method is sufficiently implemented there should be no issue with this. To help here I realise it may be useful to have a method that hides having to directly touch the dataset in the data_manager. I can add another method called `update_target(ind, target_value)` which does this bit safely.

This way the required `update_dataset` method for creating will involve just the obtaining label part for whatever process/etc it is to interface and upon obtaining the target values will just call the `update_target(ind, target_value)` method.

File: pyrelational/strategies/classification/generic_classification_strategy.py [hunk 5]
```diff
@@ -12,15 +12,15 @@ class GenericClassificationStrategy(GenericActiveLearningStrategy, ABC):
...
 
-    def active_learning_step(self, num_annotate: int):
-        self.model.train(self.l_loader, self.valid_loader)
-        output = self.model(self.u_loader).mean(0)
+    def active_learning_step(self, num_annotate: int, data_manager: GenericDataManager, model: GenericModel):
+        model.train(data_manager.get_labelled_loader(), data_manager.get_validation_loader())
+        output = model(data_manager.get_unlabelled_loader()).mean(0)
         if not torch.allclose(output.sum(1), torch.tensor(1.0)):
             output = softmax(output)
         uncertainty = self.scoring_fn(softmax(output))
```
[the remark this answers is not in the document]
Comment (paulmorio) [reply pyrelational/strategies/classification/generic_classification_strategy.py:24]: Huh. I suppose its just about removing the second one on line 24?

File: pyrelational/strategies/regression/bald_strategy.py (hunk 2 above)
Comment (paulmorio) [reply pyrelational/strategies/regression/bald_strategy.py:39]: I think we will need to chat about this at more depth as not all methods require the data_manager or model as part of the ative_learning_step.

File: pyrelational/strategies/generic_al_strategy.py [hunk 6]
```diff
@@ -1,293 +1,72 @@
...
+        # type VAR_POSITIONAL (*args) and VAR_KEYWORD (**kwargs)
+        _params = (inspect.Parameter.VAR_POSITIONAL, inspect.Parameter.VAR_KEYWORD)
+        _sign_params = inspect.signature(self.active_learning_step).parameters
+        filtered_kwargs = {
+            k: v for k, v in kwargs.items() if (k in _sign_params.keys() and _sign_params[k].kind not in _params)
+        }
+
+        exists_var_keyword = any(v.kind == inspect.Parameter.VAR_KEYWORD for v in _sign_params.values())
+        # if no kwargs filtered, return all kwargs as default
+        if not filtered_kwargs and not exists_var_keyword:
```
Comment (alicerdv) [inline pyrelational/strategies/generic_al_strategy.py:57]: If this is true, then filtered_kwargs is already an empty dictionary, so the subsequent line(`filtered_kwargs = {}`) seems redundant to me?

File: pyrelational/pipeline/generic_pipeline.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (thomasgaudelet) [reply pyrelational/pipeline/generic_pipeline.py:217]: It may be useful to rename the method as well. Something that makes it clear it's updating data manager based on current run. update_dataset_from_query, or something ?

File: pyrelational/pipeline/generic_pipeline.py [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+        """
+        if self.model.current_model is None:  # no AL steps taken so far
+            self.model.train(self.l_loader, self.valid_loader)
+
+        if test_loader is None:
+            result = self.model.test(self.test_loader)
+        else:
+            result = self.model.test(test_loader)
+        # reset current model, to avoid issues when model does not need to be trained during al_step.
+        self.model.current_model = None
```
Comment (thomasgaudelet) [inline pyrelational/pipeline/generic_pipeline.py:116]: we should probably have a method on model side doing this. It looks really dirty ...

File: pyrelational/pipeline/generic_pipeline.py [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+                or there is a new test set.
+
+        :param query: List of indices selected for labelling. Used for calculating hit ratio metric
+        :return: dictionary containing metric results on test set
+
+        """
+        if self.model.current_model is None:  # no AL steps taken so far
+            self.model.train(self.l_loader, self.valid_loader)
+
+        if test_loader is None:
```
Comment (thomasgaudelet) [inline pyrelational/pipeline/generic_pipeline.py:111]:
```suggestion
        result = self.model.test(self.test_loader if test_loader is None else test_loader)
```
would that work?

File: pyrelational/pipeline/generic_pipeline.py [hunk 9]
```diff
@@ -0,0 +1,321 @@
...
+        else:
+            result = self.model.test(test_loader)
+        # reset current model, to avoid issues when model does not need to be trained during al_step.
+        self.model.current_model = None
+
+        if self.data_manager.top_unlabelled is not None:
+            result["hit_ratio"] = (
+                np.nan
+                if query is None
+                else len(set(query) & self.data_manager.top_unlabelled) / len(self.data_manager.top_unlabelled)
```
Comment (thomasgaudelet) [inline pyrelational/pipeline/generic_pipeline.py:122]: maybe this should be done by a separate method to improve readability.

**Review by thomasgaudelet — COMMENTED** [review]
a few more comments, some might be outside scope of this PR though, just raising awareness. The unit tests should probably be refactored as well. But let's do that after merge.

File: pyrelational/pipeline/generic_pipeline.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (paulmorio) [reply pyrelational/pipeline/generic_pipeline.py:184]: It is now part of #38

File: pyrelational/strategies/regression/bald_strategy.py (hunk 2 above)
Comment (paulmorio) [reply pyrelational/strategies/regression/bald_strategy.py:39]: In our last meeting we discussed to pack the training of the model and obtaining uncertainties into a method that gets called for strategies that use this.

File: pyrelational/strategies/generic_al_strategy.py (hunk 6 above)
Comment (paulmorio) [reply pyrelational/strategies/generic_al_strategy.py]: Updated in 79dff74

File: pyrelational/pipeline/generic_pipeline.py (hunk 7 above)
Comment (paulmorio) [reply pyrelational/pipeline/generic_pipeline.py:116]: I don't remember exactly why this is here. Removing it does not affect any of the tests.

File: pyrelational/pipeline/generic_pipeline.py [hunk 10]
```diff
@@ -0,0 +1,321 @@
...
+                test data compatible with model, optional as often the test loader can be
+                generated from data_manager but is here for case when it hasn't been defined
+                or there is a new test set.
+
+        :return: performances
+        """
+        self.model.train(self.train_loader, self.valid_loader)
+
+        # use test loader in data_manager if there is one
+        result = self.model.test(self.test_loader if test_loader is None else test_loader)
```
[the remark this answers is not in the document]
Comment (paulmorio) [reply pyrelational/pipeline/generic_pipeline.py:88]: As far as I can see there is only one line in common and the methods have distinct responsibilities and meanings for the user, so I will keep them as is. I did take on Thomas' suggestion on shortening the conditional.

File: pyrelational/pipeline/generic_pipeline.py (hunk 9 above)
Comment (paulmorio) [reply pyrelational/pipeline/generic_pipeline.py:128]: Done

File: pyrelational/pipeline/generic_pipeline.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (paulmorio) [reply pyrelational/pipeline/generic_pipeline.py:88]: Done

File: pyrelational/pipeline/generic_pipeline.py [hunk 11]
```diff
@@ -0,0 +1,321 @@
...
+
+
+class GenericPipeline(ABC):
+    """The pipeline facillitates the communication between
+    - DataManager
+    - Model,
+    - ALStrategy,
+    - Oracle (Optional)
+
+    To enact a generic ative learning cycle.
```
[the remark this answers is not in the document]
Comment (paulmorio) [reply pyrelational/pipeline/generic_pipeline.py:30]: Thanks for noting them :)

File: tests/strategies/test_regression_strategies.py [hunk 12]
```diff
@@ -14,18 +15,21 @@
 def test_regression_strategies():
     model = get_model()
     data_manager = get_regression_dataset()
+    lcs = LeastConfidenceStrategy()
```
[the remark this answers is not in the document]
Comment (alicerdv) [reply tests/strategies/test_regression_strategies.py:18]: done

File: pyrelational/pipeline/generic_pipeline.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (paulmorio) [reply pyrelational/pipeline/generic_pipeline.py]: Updated interface in b0d7a61

File: pyrelational/strategies/regression/bald_strategy.py (hunk 2 above)
Comment (paulmorio) [reply pyrelational/strategies/regression/bald_strategy.py]: Updated in b81c244, however I feel that this is starting to abstract away simple one/two liners which may be confusing on its own right.

**Comment by paulmorio** [comment]
Updated and ready for another review

File: pyrelational/oracle/generic_oracle.py [hunk 13]
```diff
@@ -16,10 +16,59 @@ class GenericOracle(ABC):
...
+
+        Note this does not change the target values of the indices, this is handled by a method
+        in the oracle.
+
+        :param data_manager: reference to the data_manager whose sets we are adjusting
+        :param indices: list of indices selected for labelling
+        """
+        data_manager.update_train_labels(indices)
+
+    # @abstractmethod
```
Comment (thomasgaudelet) [inline pyrelational/oracle/generic_oracle.py:52]: should remove this if unnecessary.

File: pyrelational/oracle/generic_oracle.py [hunk 14]
```diff
@@ -16,10 +16,59 @@ class GenericOracle(ABC):
...
+    def update_target_value(self, data_manager: GenericDataManager, idx: int, value: Any) -> None:
+        """Update the target value for the observation denoted by the index
+
+        :param data_manager: reference to the data_manager whose dataset we want to update
+        :param idx: index to the observation we want to update
+        :param value: value to update the observation with
+        """
+        data_manager.set_target_value(idx=idx, value=value)
+
+    def update_target_values(self, data_manager: GenericDataManager, indices: List[int], values: List[Any]) -> None:
```
[the remark this answers is not in the document]
Comment (thomasgaudelet) [reply pyrelational/oracle/generic_oracle.py:28]: it's not the only place. But maybe that should be moved to issues and be tackled with mypy., wdyt?

File: examples/demo/scikit_estimator.py [hunk 15]
```diff
@@ -86,16 +87,19 @@ def __call__(self, loader):
 # Instantiate an active learning strategy
 al_strategy = LeastConfidenceStrategy()
 
+# Instantiate an oracle (in this case a dummy one)
+oracle = DummyOracle()
+
 # Given that we have a data manager, a model, and an active learning strategy
 # we may create an active learning pipeline
-pipeline = GenericPipeline(data_manager=data_manager, model=model, strategy=al_strategy)
+pipeline = GenericPipeline(data_manager=data_manager, model=model, strategy=al_strategy, oracle=oracle)
```
Comment (thomasgaudelet) [inline examples/demo/scikit_estimator.py:95]: should we have DummyOracle by default in GenericPipeline?

Comment (thomasgaudelet) [reply examples/demo/scikit_estimator.py:95]: also should we rename to BenchmarkOracle? DefaultOracle?

**Review by thomasgaudelet — COMMENTED** [review]
looks good! small comments, some maybe should be moved to issues instead of being addressed in this PR.

File: pyrelational/pipeline/generic_pipeline.py [hunk 16]
```diff
@@ -0,0 +1,312 @@
...
+        oracle: Optional[GenericOracle] = None,
+    ):
+        super(GenericPipeline, self).__init__()
+        self.data_manager = data_manager
+        self.model = model
+        self.strategy = strategy
+        self.oracle = oracle
+
+        # Mediator pattern for strategy
+        self.strategy._pipeline = self
```
Comment (thomasgaudelet) [inline pyrelational/pipeline/generic_pipeline.py:59]: does this need to be removed in the end? bit confused

File: pyrelational/pipeline/generic_pipeline.py [hunk 17]
```diff
@@ -0,0 +1,312 @@
...
+        and theoretical performance methods.
+
+        :param result: Dict or Dict-like of metrics
+        :param query: List of indices selected for labelling. Usef for calculating hit ratio metric
+        """
+        if self.data_manager.top_unlabelled is not None:
+            result["hit_ratio"] = (
+                np.nan
+                if query is None
+                else len(set(query) & self.data_manager.top_unlabelled) / len(self.data_manager.top_unlabelled)
```
Comment (thomasgaudelet) [inline pyrelational/pipeline/generic_pipeline.py:128]: has this been moved to a separate function in a different PR?

File: pyrelational/pipeline/generic_pipeline.py [hunk 18]
```diff
@@ -0,0 +1,312 @@
...
+        """
+        self.model.train(self.train_loader, self.valid_loader)
+
+        # use test loader in data_manager if there is one
+        result = self.model.test(self.test_loader if test_loader is None else test_loader)
+        result = self.compute_hit_ratio(result)
+        self.performances["full"] = result
+
+        # make sure that theoretical best model is not stored
+        self.model.current_model = None
```
Comment (alicerdv) [inline pyrelational/pipeline/generic_pipeline.py:93]: if self.model.current_model = None has been removed in current_performance, does it also need to be removed here?

Comment (alicerdv) [reply pyrelational/pipeline/generic_pipeline.py:93]: actually, I guess this serves a different purpose in this case, so here it is necessary?

File: examples/demo/scikit_estimator.py (hunk 15 above)
Comment (paulmorio) [reply examples/demo/scikit_estimator.py:95]: I think BenchmarkOracle is most indicative about its function. Will change it to that.

File: pyrelational/pipeline/generic_pipeline.py (hunk 16 above)
Comment (paulmorio) [reply pyrelational/pipeline/generic_pipeline.py:59]: Thats weird I thought I removed it explicitly in an earlier commit. Thanks for catching it.

File: pyrelational/oracle/generic_oracle.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (thomasgaudelet) [reply pyrelational/oracle/generic_oracle.py:28]: Created issue #41 to this end.

File: pyrelational/oracle/generic_oracle.py (hunk 13 above)
Comment (paulmorio) [reply pyrelational/oracle/generic_oracle.py:52]: Why should this be removed? It is used in the `update_dataset` method in the oracle, and is no longer a method in the pipeline. I can disencourage access to it by making it private.

File: pyrelational/pipeline/generic_pipeline.py (hunk 18 above)
Comment (thomasgaudelet) [reply pyrelational/pipeline/generic_pipeline.py:93]: Yes, we should be careful about changing these things without proper checks and it does fall outside of the scope of this PR.

We should however go through repo and make sure that we don't have these kind of attribute setting calls outside of a class. e.g. this should be replaced by a reset_current_model() method in the ModelManager.

File: pyrelational/oracle/generic_oracle.py (hunk 13 above)
Comment (thomasgaudelet) [reply pyrelational/oracle/generic_oracle.py:52]: because it's commented out :p. If we need it, then it should be uncommented!

File: pyrelational/pipeline/generic_pipeline.py (hunk 17 above)
Comment (paulmorio) [reply pyrelational/pipeline/generic_pipeline.py:128]: No this was one of the points you and Alice raised about the 4 lines being repeated in `theoretical_performance` and `current_performance` so it is now its own method called `compute_hit_ratio`

Comment (thomasgaudelet) [reply pyrelational/pipeline/generic_pipeline.py:128]: ah, right, makes sense! I think one of my comment was about hiding this specific line behind a function name, but this makes more sense !

File: pyrelational/pipeline/generic_pipeline.py (hunk 18 above)
Comment (paulmorio) [reply pyrelational/pipeline/generic_pipeline.py:93]: It is necessary as we don't want to inadvertedly use the model trained on the full training data during any stage of the active learning cycle and give overly optimistic results.

It does raise another question about the situation if someone calls the theoretical_performance at some stage (though I don't see why), it will wipe the currently saved model in the pipeline.

File: pyrelational/oracle/generic_oracle.py (hunk 13 above)
Comment (paulmorio) [reply pyrelational/oracle/generic_oracle.py:52]: Oh I see you meant the decorator!, sorry I thought the window is referring to the `update_annotations` method. Yes the decorator is now uncommented.

Comment (paulmorio) [reply pyrelational/oracle/generic_oracle.py]: Is there a way to expand the view or get a more direct highlight of the line in question in this review mode?

File: examples/demo/scikit_estimator.py (hunk 15 above)
Comment (paulmorio) [reply examples/demo/scikit_estimator.py:95]: Done

File: pyrelational/oracle/generic_oracle.py (hunk 13 above)
Comment (paulmorio) [reply pyrelational/oracle/generic_oracle.py]: Done

File: pyrelational/pipeline/generic_pipeline.py (hunk 16 above)
Comment (paulmorio) [reply pyrelational/pipeline/generic_pipeline.py]: Done

File: pyrelational/data/data_manager.py [hunk 19]
```diff
@@ -217,7 +217,10 @@ def set_target_value(self, idx: int, value: Any) -> None:
         :param idx: index value to the observation
         :param value: new value for the observation
         """
-        self.dataset.y[idx] = value
+        if hasattr(self.dataset, "y"):
```
Comment (thomasgaudelet) [inline pyrelational/data/data_manager.py:220]: didn't know this was an issue :(, maybe we should have a attr_name input to data_manager that defaults to "y" or something? so users can adapt it

Comment (paulmorio) [reply pyrelational/data/data_manager.py:220]: Sure thing, I was retrofitting it here for some of the default datasets we have like the MNIST ones in the examples which uses the target attribute

Comment (thomasgaudelet) [reply pyrelational/data/data_manager.py:220]: Ok, didn't see it was moved to a new issue!

**Review by thomasgaudelet — APPROVED** [review]

File: pyrelational/pipeline/generic_pipeline.py (hunk 18 above)
Comment (alicerdv) [reply pyrelational/pipeline/generic_pipeline.py:95]: okay, moved what Thomas mentioned into a new issue

**Review by alicerdv — APPROVED** [review]

**Review by a-pouplin — APPROVED** [review]
